Getting `make develop` to work with Clang and different GCC versions

As I mentioned in #281, I tried running `make develop` for the first time recently. However, the current set of options don't seem to be compatible with Clang (specifically, `Apple LLVM version 9.1.0` included with Xcode 9.4), so I started looking into ways to make it compatible.

In addition, when trying to run it on my Ubuntu 16.04 VM, neither the default GCC 5.5, nor the more recent GCC 8.0 appeared to support the options in `make develop`, either.

This took me down a bit of a rabbit hole, in terms of trying to find what options were needed to get it working, and what options were compatible with both. I then realised that some discussion might be need about how best to proceed with this.

## Questions
- Should we continue to use `-Wpedantic`, knowing that the exact warnings may change between compiler versions?
- If so, do we limit our options to warnings that are supported by Clang and most versions of GCC? This would make things easy, but would make the compile-time checking less thorough.
- If no to either of the above, do we give different warning flags to each compiler, to take advantages of the warnings that both are able to provide? I'm not sure what the best way of doing this in `make` would be.

While I don't mind if we use `-Wpedantic` or not, I would personally prefer to enable specific warnings for each compiler, so we get the advice of both compilers when running on Travis CI, although I can see why hard-coded compiler checks may be undesirable.

In addition, I would also suggest:
- Use `make develop` in Travis CI builds, to ensure that incoming PRs don't trigger any new warnings, and to ensure that the flags are compatible with the compilers that are available there.
    - While we're there, I would advocate re-enabling the Mac builds on Travis CI; I think they've got past their blip now.

## More detailed changes
These are the specific changes that I had to make in order to use `make develop` on each of these compilers.

### Clang
It requires the following flag changes:
```diff
- -Wchkp
- -Wformat-overflow=2
- -Wformat-truncation=1
- -Wstringop-overflow=4
- -Walloc-zero
- -Wduplicated-cond
- -Wlogical-op
- -Wno-aggressive-loop-optimizations
+ -O1                                (needed for -fsanitize=object-size; GCC also recommends optimisations)
+ -Wno-missing-field-initializers    (needed for -Wpedantic; GCC-compatible)
+ -Wno-format-nonliteral             (needed for -Wpedantic; GCC-compatible)
```

Granted, I've not tried the latest open-source LLVM build, so that may well support some of these flags, unlike the version of LLVM that is included with Xcode.

It also flags a genuine warning in `src/link/object.c`:
```
src/link/object.c:286:14: warning: cast from 'uint8_t *' (aka 'unsigned char *')
      to 'struct sSymbol **' increases required alignment from 1 to 8
      [-Wcast-align]
                tSymbols = (struct sSymbol **)&dummymem;
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
```

I'm not entirely sure why we need `dummymem` – I'd have thought removing it and simply using `NULL` would be fine, and it does appear to work on the test repos, but I've not double-checked the code any further.

*Aside: We could also take advantage of Clang's runtime sanitisers, to catch any undefined behaviour, memory problems, etc. while running the test repos. Particularly something to think about while running things over Travis!* 🙂 

### GCC 5.5
It requires the following flag changes:
```diff
- -Wformat-overflow=2
- -Wformat-truncation=1
- -Wstringop-overflow=4
- -Walloc-zero
- -Wduplicated-cond
```

(Though apparently I've not configured it properly, so I can't get it to link 😛)

### GCC 8.0
It requires the following flag changes:
```diff
+ -Wno-strict-overflow    (needed for -Wpedantic; Clang-compatible)
```

(Again, my VM seems to be borked on the linking front)

…And that's enough of a wall of text 😆 If you need more help with figuring out what does and doesn't work, just let me know and I'll be happy to give things a try.

--- a/.github/workflows/testing.yml
+++ b/.github/workflows/testing.yml
@@ -16,19 +16,6 @@ jobs:
             cc: gcc
           - os: macos-11.0
             cc: gcc
-        include:
-          - os: ubuntu-18.04
-            target: develop
-            cmakevars: -DSANITIZERS=ON -DMORE_WARNINGS=ON -DCMAKE_BUILD_TYPE=Debug
-          - os: ubuntu-20.04
-            target: develop
-            cmakevars: -DSANITIZERS=ON -DMORE_WARNINGS=ON -DCMAKE_BUILD_TYPE=Debug
-          - os: macos-11.0
-            target: develop
-            cmakevars: -DSANITIZERS=ON -DMORE_WARNINGS=ON -DCMAKE_BUILD_TYPE=Debug
-          - os: macos-10.15
-            target: develop
-            cmakevars: -DSANITIZERS=ON -DMORE_WARNINGS=ON -DCMAKE_BUILD_TYPE=Debug
       fail-fast: false
     runs-on: ${{ matrix.os }}
     steps:
@@ -43,13 +30,13 @@ jobs:
       - name: Build & install using Make
         run: |
           export PATH="/usr/local/opt/bison/bin:$PATH"
-          make ${{ matrix.target }} -j Q= CC=${{ matrix.cc }}
+          make develop -j Q= CC=${{ matrix.cc }}
           sudo make install -j Q=
         if: matrix.buildsys == 'make'
       - name: Build & install using CMake
         run: |
           export PATH="/usr/local/opt/bison/bin:$PATH"
-          cmake -S . -B build -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=${{ matrix.cc }} ${{ matrix.cmakevars }}
+          cmake -S . -B build -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_BUILD_TYPE=Debug -DCMAKE_C_COMPILER=${{ matrix.cc }} -DSANITIZERS=ON -DMORE_WARNINGS=ON
           cmake --build build -j
           cp build/src/rgb{asm,link,fix,gfx} .
           sudo cmake --install build

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -45,15 +45,19 @@ else()
   endif()
 
   if(MORE_WARNINGS)
-    add_compile_options(-Werror -Wextra -Wno-type-limits
-                        -Wno-sign-compare -Wvla -Wformat -Wformat-security -Wformat-overflow=2
-                        -Wformat-truncation=1 -Wformat-y2k -Wswitch-enum -Wunused
-                        -Wuninitialized -Wunknown-pragmas -Wstrict-overflow=5
-                        -Wstringop-overflow=4 -Walloc-zero -Wduplicated-cond
-                        -Wfloat-equal -Wshadow -Wcast-qual -Wcast-align -Wlogical-op
-                        -Wnested-externs -Wno-aggressive-loop-optimizations -Winline
-                        -Wundef -Wstrict-prototypes -Wold-style-definition
-                        -Wno-unknown-warning-option -Wno-tautological-constant-out-of-range-compare)
+    add_compile_options(-Werror -Wextra
+                        -Walloc-zero -Wcast-align -Wcast-qual -Wduplicated-cond
+                        -Wfloat-equal -Winline -Wlogical-op -Wnested-externs -Wold-style-definition
+                        -Wstrict-overflow=5 -Wstrict-prototypes -Wundef -Wuninitialized -Wunused
+                        -Wshadow # TODO: -Wshadow=compatible-local ?
+                        -Wstringop-overflow=4 # TODO: would work better with optimizations
+                        -Wno-sign-compare # TODO: fix those warnings
+                        -Wformat=2 -Wformat-overflow=2 -Wformat-truncation=1
+                        -Wno-format-nonliteral # We have a couple of "dynamic" prints
+                        # We do some range checks that are always false on some platforms (GCC, Clang)
+                        -Wno-type-limits -Wno-tautological-constant-out-of-range-compare
+                        -Wvla # MSVC does not support VLAs
+                        -Wno-unknown-warning-option) # Clang shouldn't diagnose unknown warnings
   endif()
 endif()
 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -46,11 +46,12 @@ else()
 
   if(MORE_WARNINGS)
     add_compile_options(-Werror -Wextra
-                        -Walloc-zero -Wcast-align -Wcast-qual -Wduplicated-cond
+                        -Walloc-zero -Wcast-align -Wcast-qual -Wduplicated-branches -Wduplicated-cond
                         -Wfloat-equal -Winline -Wlogical-op -Wnested-externs -Wold-style-definition
+                        -Wshift-overflow=2
                         -Wstrict-overflow=5 -Wstrict-prototypes -Wundef -Wuninitialized -Wunused
                         -Wshadow # TODO: -Wshadow=compatible-local ?
-                        -Wstringop-overflow=4 # TODO: would work better with optimizations
+                        -Wnull-dereference -Wstringop-overflow=4 # TODO: would work better with optimizations
                         -Wno-sign-compare # TODO: fix those warnings
                         -Wformat=2 -Wformat-overflow=2 -Wformat-truncation=1
                         -Wno-format-nonliteral # We have a couple of "dynamic" prints

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -54,7 +54,6 @@ else()
                         -Wold-style-definition -Wshift-overflow=2 -Wstrict-overflow=5
                         -Wstrict-prototypes -Wstringop-overflow=4 -Wundef -Wuninitialized -Wunused
                         -Wshadow # TODO: -Wshadow=compatible-local ?
-                        -Wno-sign-compare # TODO: fix those warnings
                         -Wformat=2 -Wformat-overflow=2 -Wformat-truncation=1
                         -Wno-format-nonliteral # We have a couple of "dynamic" prints
                         # We do some range checks that are always false on some platforms (GCC, Clang)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -42,16 +42,18 @@ else()
                   -fsanitize=alignment -fsanitize=null -fsanitize=address)
     add_compile_options(${SAN_FLAGS})
     link_libraries(${SAN_FLAGS})
+    # A non-zero optimization level is desired in debug mode, but allow overriding it nonetheless
+    # TODO: this overrides anything previously set... that's a bit sloppy!
+    set(CMAKE_C_FLAGS_DEBUG "-g -Og -fno-omit-frame-pointer -fno-optimize-sibling-calls" CACHE STRING "" FORCE)
   endif()
 
   if(MORE_WARNINGS)
     add_compile_options(-Werror -Wextra
                         -Walloc-zero -Wcast-align -Wcast-qual -Wduplicated-branches -Wduplicated-cond
-                        -Wfloat-equal -Winline -Wlogical-op -Wnested-externs -Wold-style-definition
-                        -Wshift-overflow=2
-                        -Wstrict-overflow=5 -Wstrict-prototypes -Wundef -Wuninitialized -Wunused
+                        -Wfloat-equal -Winline -Wlogical-op -Wnested-externs -Wnull-dereference
+                        -Wold-style-definition -Wshift-overflow=2 -Wstrict-overflow=5
+                        -Wstrict-prototypes -Wstringop-overflow=4 -Wundef -Wuninitialized -Wunused
                         -Wshadow # TODO: -Wshadow=compatible-local ?
-                        -Wnull-dereference -Wstringop-overflow=4 # TODO: would work better with optimizations
                         -Wno-sign-compare # TODO: fix those warnings
                         -Wformat=2 -Wformat-overflow=2 -Wformat-truncation=1
                         -Wno-format-nonliteral # We have a couple of "dynamic" prints

--- a/Makefile
+++ b/Makefile
@@ -209,17 +209,21 @@ checkdiff:
 # This target is used during development in order to prevent adding new issues
 # to the source code. All warnings are treated as errors in order to block the
 # compilation and make the continous integration infrastructure return failure.
+# The rationale for some of the flags is documented in the CMakeLists.
 
 develop:
-	$Qenv ${MAKE} WARNFLAGS="-Werror -Wall -Wextra -Wpedantic -Wno-type-limits \
-		-Wno-sign-compare -Wvla -Wformat -Wformat-security -Wformat-overflow=2 \
-		-Wformat-truncation=1 -Wformat-y2k -Wswitch-enum -Wunused \
-		-Wuninitialized -Wunknown-pragmas -Wstrict-overflow=4 \
-		-Wstringop-overflow=4 -Walloc-zero -Wduplicated-cond \
-		-Wfloat-equal -Wshadow -Wcast-qual -Wcast-align -Wlogical-op \
-		-Wnested-externs -Wno-aggressive-loop-optimizations -Winline \
-		-Wundef -Wstrict-prototypes -Wold-style-definition \
-		-Wno-unknown-warning-option -Wno-tautological-constant-out-of-range-compare \
+	$Qenv ${MAKE} WARNFLAGS="-Werror -Wextra \
+		-Walloc-zero -Wcast-align -Wcast-qual -Wduplicated-cond \
+		-Wfloat-equal -Winline -Wlogical-op -Wnested-externs -Wold-style-definition \
+		-Wstrict-overflow=5 -Wstrict-prototypes -Wundef -Wuninitialized -Wunused \
+		-Wshadow \
+		-Wstringop-overflow=4 \
+		-Wno-sign-compare \
+		-Wformat=2 -Wformat-overflow=2 -Wformat-truncation=1 \
+		-Wno-format-nonliteral \
+		-Wno-type-limits -Wno-tautological-constant-out-of-range-compare \
+		-Wvla \
+		-Wno-unknown-warning-option \
 		-fsanitize=shift -fsanitize=integer-divide-by-zero \
 		-fsanitize=unreachable -fsanitize=vla-bound \
 		-fsanitize=signed-integer-overflow -fsanitize=bounds \

--- a/Makefile
+++ b/Makefile
@@ -213,11 +213,12 @@ checkdiff:
 
 develop:
 	$Qenv ${MAKE} WARNFLAGS="-Werror -Wextra \
-		-Walloc-zero -Wcast-align -Wcast-qual -Wduplicated-cond \
+		-Walloc-zero -Wcast-align -Wcast-qual -Wduplicated-branches -Wduplicated-cond \
 		-Wfloat-equal -Winline -Wlogical-op -Wnested-externs -Wold-style-definition \
+		-Wshift-overflow=2 \
 		-Wstrict-overflow=5 -Wstrict-prototypes -Wundef -Wuninitialized -Wunused \
 		-Wshadow \
-		-Wstringop-overflow=4 \
+		-Wnull-dereference -Wstringop-overflow=4 \
 		-Wno-sign-compare \
 		-Wformat=2 -Wformat-overflow=2 -Wformat-truncation=1 \
 		-Wno-format-nonliteral \

--- a/Makefile
+++ b/Makefile
@@ -230,7 +230,7 @@ develop:
 		-fsanitize=signed-integer-overflow -fsanitize=bounds \
 		-fsanitize=object-size -fsanitize=bool -fsanitize=enum \
 		-fsanitize=alignment -fsanitize=null -fsanitize=address" \
-		CFLAGS="-ggdb3 -O0"
+		CFLAGS="-ggdb3 -Og -fno-omit-frame-pointer -fno-optimize-sibling-calls"
 
 # Targets for the project maintainer to easily create Windows exes.
 # This is not for Windows users!

--- a/Makefile
+++ b/Makefile
@@ -30,7 +30,7 @@ PNGLDLIBS	:= `${PKG_CONFIG} --libs-only-l libpng`
 # Note: if this comes up empty, `version.c` will automatically fall back to last release number
 VERSION_STRING	:= `git describe --tags --dirty --always 2>/dev/null`
 
-WARNFLAGS	:= -Wall
+WARNFLAGS	:= -Wall -pedantic
 
 # Overridable CFLAGS
 CFLAGS		?= -O3 -flto -DNDEBUG

--- a/Makefile
+++ b/Makefile
@@ -219,7 +219,6 @@ develop:
 		-Wstrict-overflow=5 -Wstrict-prototypes -Wundef -Wuninitialized -Wunused \
 		-Wshadow \
 		-Wnull-dereference -Wstringop-overflow=4 \
-		-Wno-sign-compare \
 		-Wformat=2 -Wformat-overflow=2 -Wformat-truncation=1 \
 		-Wno-format-nonliteral \
 		-Wno-type-limits -Wno-tautological-constant-out-of-range-compare \

--- a/Makefile
+++ b/Makefile
@@ -211,7 +211,7 @@ checkdiff:
 # compilation and make the continous integration infrastructure return failure.
 
 develop:
-	$Qenv $(MAKE) WARNFLAGS="-Werror -Wall -Wextra -Wpedantic -Wno-type-limits \
+	$Qenv ${MAKE} WARNFLAGS="-Werror -Wall -Wextra -Wpedantic -Wno-type-limits \
 		-Wno-sign-compare -Wvla -Wformat -Wformat-security -Wformat-overflow=2 \
 		-Wformat-truncation=1 -Wformat-y2k -Wswitch-enum -Wunused \
 		-Wuninitialized -Wunknown-pragmas -Wstrict-overflow=4 \
@@ -233,11 +233,11 @@ develop:
 # install instructions instead.
 
 mingw32:
-	$Qmake CC=i686-w64-mingw32-gcc BISON=bison \
+	$Q${MAKE} CC=i686-w64-mingw32-gcc BISON=bison \
 		PKG_CONFIG=i686-w64-mingw32-pkg-config -j
 
 mingw64:
-	$Qmake CC=x86_64-w64-mingw32-gcc BISON=bison \
+	$Q${MAKE} CC=x86_64-w64-mingw32-gcc BISON=bison \
 		PKG_CONFIG=x86_64-w64-mingw32-pkg-config -j
 
 wine-shim:

--- a/include/asm/macro.h
+++ b/include/asm/macro.h
@@ -25,7 +25,7 @@ void macro_AppendArg(struct MacroArgs **args, char *s);
 void macro_UseNewArgs(struct MacroArgs *args);
 void macro_FreeArgs(struct MacroArgs *args);
 char const *macro_GetArg(uint32_t i);
-char *macro_GetAllArgs(void);
+char const *macro_GetAllArgs(void);
 
 uint32_t macro_GetUniqueID(void);
 char const *macro_GetUniqueIDStr(void);

--- a/include/asm/opt.h
+++ b/include/asm/opt.h
@@ -11,8 +11,8 @@
 
 #include <stdbool.h>
 
-void opt_B(char chars[2]);
-void opt_G(char chars[4]);
+void opt_B(char const chars[2]);
+void opt_G(char const chars[4]);
 void opt_P(uint8_t fill);
 void opt_L(bool optimize);
 void opt_W(char const *flag);

--- a/include/gfx/main.h
+++ b/include/gfx/main.h
@@ -24,7 +24,7 @@ struct Options {
 	bool mirror;
 	bool unique;
 	bool colorcurve;
-	int trim;
+	unsigned int trim;
 	char *tilemapfile;
 	bool tilemapout;
 	char *attrmapfile;
@@ -43,7 +43,7 @@ struct RGBColor {
 
 struct ImageOptions {
 	bool horizontal;
-	int trim;
+	unsigned int trim;
 	char *tilemapfile;
 	bool tilemapout;
 	char *attrmapfile;

--- a/src/asm/macro.c
+++ b/src/asm/macro.c
@@ -101,7 +101,7 @@ char const *macro_GetArg(uint32_t i)
 					      : macroArgs->args[realIndex];
 }
 
-char *macro_GetAllArgs(void)
+char const *macro_GetAllArgs(void)
 {
 	if (!macroArgs)
 		return NULL;

--- a/src/asm/opt.c
+++ b/src/asm/opt.c
@@ -24,12 +24,12 @@ struct OptStackEntry {
 
 static struct OptStackEntry *stack = NULL;
 
-void opt_B(char chars[2])
+void opt_B(char const chars[2])
 {
 	lexer_SetBinDigits(chars);
 }
 
-void opt_G(char chars[4])
+void opt_G(char const chars[4])
 {
 	lexer_SetGfxDigits(chars);
 }

--- a/src/asm/output.c
+++ b/src/asm/output.c
@@ -192,7 +192,7 @@ static uint32_t getsectid(struct Section const *sect)
 
 static uint32_t getSectIDIfAny(struct Section const *sect)
 {
-	return sect ? getsectid(sect) : -1;
+	return sect ? getsectid(sect) : (uint32_t)-1;
 }
 
 /*
@@ -200,7 +200,7 @@ static uint32_t getSectIDIfAny(struct Section const *sect)
  */
 static void writepatch(struct Patch const *patch, FILE *f)
 {
-	assert(patch->src->ID != -1);
+	assert(patch->src->ID != (uint32_t)-1);
 	putlong(patch->src->ID, f);
 	putlong(patch->lineNo, f);
 	putlong(patch->offset, f);
@@ -264,7 +264,7 @@ static void writesymbol(struct Symbol const *sym, FILE *f)
 	if (!sym_IsDefined(sym)) {
 		putc(SYMTYPE_IMPORT, f);
 	} else {
-		assert(sym->src->ID != -1);
+		assert(sym->src->ID != (uint32_t)-1);
 
 		putc(sym->isExported ? SYMTYPE_EXPORT : SYMTYPE_LOCAL, f);
 		putlong(sym->src->ID, f);
@@ -490,7 +490,7 @@ static void freeassert(struct Assertion *assert)
 
 static void writeFileStackNode(struct FileStackNode const *node, FILE *f)
 {
-	putlong(node->parent ? node->parent->ID : -1, f);
+	putlong(node->parent ? node->parent->ID : (uint32_t)-1, f);
 	putlong(node->lineNo, f);
 	putc(node->type, f);
 	if (node->type != NODE_REPT) {

--- a/src/gfx/gb.c
+++ b/src/gfx/gb.c
@@ -37,15 +37,14 @@ void transpose_tiles(struct GBImage *gb, int width)
 
 void raw_to_gb(const struct RawIndexedImage *raw_image, struct GBImage *gb)
 {
-	int x, y, byte;
 	uint8_t index;
 
-	for (y = 0; y < raw_image->height; y++) {
-		for (x = 0; x < raw_image->width; x++) {
+	for (unsigned int y = 0; y < raw_image->height; y++) {
+		for (unsigned int x = 0; x < raw_image->width; x++) {
 			index = raw_image->data[y][x];
 			index &= (1 << depth) - 1;
 
-			byte = y * depth
+			unsigned int byte = y * depth
 				+ x / 8 * raw_image->height / 8 * 8 * depth;
 			gb->data[byte] |= (index & 1) << (7 - x % 8);
 			if (depth == 2) {

--- a/src/gfx/makepng.c
+++ b/src/gfx/makepng.c
@@ -95,6 +95,9 @@ void output_png_file(const struct Options *opts,
 	if (!f)
 		err("Opening output png file '%s' failed", outfile);
 
+	if (opts->debug)
+		free(outfile);
+
 	img.png = png_create_write_struct(PNG_LIBPNG_VER_STRING,
 					  NULL, NULL, NULL);
 	if (!img.png)
@@ -135,9 +138,6 @@ void output_png_file(const struct Options *opts,
 
 	png_destroy_write_struct(&img.png, &img.info);
 	fclose(f);
-
-	if (opts->debug)
-		free(outfile);
 }
 
 void destroy_raw_image(struct RawIndexedImage **raw_image_ptr_ptr)

--- a/src/gfx/makepng.c
+++ b/src/gfx/makepng.c
@@ -142,10 +142,9 @@ void output_png_file(const struct Options *opts,
 
 void destroy_raw_image(struct RawIndexedImage **raw_image_ptr_ptr)
 {
-	int y;
 	struct RawIndexedImage *raw_image = *raw_image_ptr_ptr;
 
-	for (y = 0; y < raw_image->height; y++)
+	for (unsigned int y = 0; y < raw_image->height; y++)
 		free(raw_image->data[y]);
 
 	free(raw_image->data);

--- a/src/link/patch.c
+++ b/src/link/patch.c
@@ -111,7 +111,7 @@ static uint32_t getRPNByte(uint8_t const **expression, int32_t *size,
 static struct Symbol const *getSymbol(struct Symbol const * const *symbolList,
 				      uint32_t index)
 {
-	assert(index != -1); /* PC needs to be handled specially, not here */
+	assert(index != (uint32_t)-1); /* PC needs to be handled specially, not here */
 	struct Symbol const *symbol = symbolList[index];
 
 	/* If the symbol is defined elsewhere... */

